Libs/vtkSegmentationCore/vtkBinaryLabelmapToClosedSurfaceConversionRule.cxx

```diff
@@ -288,39 +291,86 @@ bool vtkBinaryLabelmapToClosedSurfaceConversionRule::CreateClosedSurface(vtkOrie
   double smoothingFactor = this->ConversionParameters->GetValueAsDouble(GetSmoothingFactorParameterName());
   int computeSurfaceNormals = this->ConversionParameters->GetValueAsInt(GetComputeSurfaceNormalsParameterName());
 
-  vtkNew<vtkDiscreteFlyingEdges3D> marchingCubes;
-  marchingCubes->SetInputData(binaryLabelmapWithIdentityGeometry);
-  marchingCubes->ComputeGradientsOff();
-  marchingCubes->ComputeNormalsOff(); // While computing normals is faster using the flying edges filter,
-  // it results in incorrect normals in meshes from shared labelmaps marchingCubes->ComputeScalarsOn();
+  // Generation method
+  // 0 = flying edges
+  // 1 = surface nets
+  int generationMethod = this->ConversionParameters->GetValueAsInt(GetConversionMethodParameterName());
+  vtkNew<vtkDiscreteFlyingEdges3D> flyingEdges;
+  vtkNew<vtkSurfaceNets3D> surfaceNets;
+  vtkSmartPointer<vtkPolyData> processingResult = vtkSmartPointer<vtkPolyData>::New();
+
+  if (generationMethod == 0)
+  {
+    flyingEdges->SetInputData(binaryLabelmapWithIdentityGeometry);
+    flyingEdges->ComputeGradientsOff();
+    flyingEdges->ComputeNormalsOff(); // While computing normals is faster using the flying edges filter,
+    // it results in incorrect normals in meshes from shared labelmaps marchingCubes->ComputeScalarsOn();
+
+    int valueIndex = 0;
+    for (vtkIdType labelValue : labelValues)
+    {
+      flyingEdges->SetValue(valueIndex, labelValue);
+      ++valueIndex;
+    }
+    try
+    {
+      flyingEdges->Update();
+    }
+    catch (...)
+    {
+      vtkErrorMacro("Convert: Error while running flying edges!");
+      return false;
+    }
+    processingResult = flyingEdges->GetOutput();
+  }
+  else if (generationMethod == 1)
+  {
+    vtkSmartPointer<vtkSurfaceNets3D> surfaceNets = vtkSmartPointer<vtkSurfaceNets3D>::New();
+    surfaceNets->SetInputData(binaryLabelmapWithIdentityGeometry);
+    // Disable internal smoothing, and use vtkWindowedSincPolyDataFilter for smoothing as needed
+    surfaceNets->SmoothingOff();
```

Comment:
Doe the built-in smoothing capability provide parameter similar to the ones we are using with `vtkDecimatePro` ? 

If the bottleneck is currently related to using `vtkDecimatePro`, we should consider using the optimized method builtin `vtkSurfaceNets3D`

Reply:
From @wschroed 

> Generally the smoothing in vtkSurfaceNets3D is better. It's certainly going to be faster. It is possible to turn off smoothing in vtkSurfaceNets3D and then use an external filter like vtkWindowedSincPolyDataFilter (or other).
>
> Working with the algorithm deeply, and talking to @mhalle  and @sfriskenbwh, there are certain situations where resolution effects come into play.

Reply:
Follow-up from @wschroed 

> The reason smoothing is faster using the built-in smoothing process in `vtkSurfaceNets3D` is that the smoothing stencils are constructed as part of the algorithm. If you use an external smoother like `vtkWindowedSincPolyDataFilter`, the stencils have to be built from scratch. Also, the built-in smoothing stencils are somewhat optimized so they generally produce better results.
>
> For low-res volumes and associated smoothing, artifacts can appear. This is due to what I call "adhesions" where multiple materials all join in the same voxel cell. This results in over constraining the smoothing process, which can produce the artifacts. I believe this can be fixed by modification to the algorithm (and smoothing stencils) - it would take some serious work [...]